### PR TITLE
Init extractors/ukwiki.py

### PR DESCRIPTION
**Pull request by chtnnh**

### CODE REVIEW

File: articlequality/extractors/ukwiki.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        except ValueError as e:
+            logger.warning(traceback.format_exc())
+            pass  # no assessment class in template
+
+
+def get_quality(template):
+    LABEL_MAP = {observed_label: normalized_label
+             for normalized_label, observed_labels in 
+             NORMALIZED_LABELS.items()
+             for observed_label in observed_labels}
```
Comment (halfak) [inline articlequality/extractors/ukwiki.py:43]: You want this LABEL_MAP at the top so that it is only generated once.  Then you'll use the LABEL_MAP here to check to see if a label from a "class" or "рівень" matches an expected string (keys of LABEL_MAP) and to convert it to the canonical form (values of LABEL_MAP).

File: articlequality/extractors/ukwiki.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+def normalize_template_name(template_name):
+    template_name = str(template_name).lower().replace("_", " ")
+
+
+def normalize_project_name(template_name):
+    return WP_PREFIX.sub('', template_name.lower().replace("_", " ")).strip()
```
Comment (halfak) [inline articlequality/extractors/ukwiki.py:50]: We'll need to define:
```
WP_PREFIX = re.compile(r"стаття про[еє]кту|вікіпро[еє]кт|про[еє]кт")
```

File: articlequality/extractors/ukwiki.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        "I": ["I", "1"],
+        "II": ["II", "2"],
+        "III": ["III", "3"],
+        "IV": ["IV", "4", "Stub", "stub"]
+}
+
+
+def from_template(template):
+
+    if template.has_param('class') or template.has_param('рівень'):
```
Comment (halfak) [inline articlequality/extractors/ukwiki.py:21]: I'm thinking we should change this pattern.  It should be something like:
```
template_name = normalize_template_name(template.name)

if WP_PREFIX.match(template_name):
    project_name = normalize_project_name(template_name)
    label = get_quality_label(template)
    if label is None:
        logger.warn("Couldn't extract label from {0}".format(template)
    else:
        yield (project_name, label)
```

File: articlequality/extractors/ukwiki.py (hunk 1 above)
Comment (halfak) [reply articlequality/extractors/ukwiki.py:43]:
```
if template.has_param("рівень") or template.has_param("class"):
  if template.has_param("class"): 
    label = template.get_param("рівень")
  else:
    label = template.has_param("class")
  return LABEL_MAP.get(label)
else:
  return None
```

File: articlequality/extractors/ukwiki.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            label = template.get_param("рівень")
+        else:
+            label = template.has_param("class")
+        return LABEL_MAP.get(label)
+    else:
+        return None
+
+
+def normalize_template_name(template_name):
+    template_name = str(template_name).lower().replace("_", " ")
```
Comment (halfak) [inline articlequality/extractors/ukwiki.py:49]: `return str(template_name).lower().replace("_", " ")`

File: Makefile [hunk 5]
```diff
@@ -836,6 +836,13 @@ trwiki_models: \
 trwiki_tuning_reports: \
 	tuning_reports/trwiki.wp10.md
 
+####################### Ukrainian Wikipedia #################################
+datasets/ukwiki.labelings.20200501.json:
+	./utility extract_labelings \
+		/mnt/data/xmldatadumps/public/ukwiki/20200501/ukwiki-20200501-pages-meta-history?*.xml-*.bz2 \
```
Comment (he7d3r) [inline Makefile:842]: Is the `?` after `history` a typo?

Comment (halfak) [reply Makefile:842]: It ensure that we match `ukwiki-20200501-pages-meta-history1.xml-*.bz2` but that we don't match `ukwiki-20200501-pages-meta-history.xml-*.bz2`.  The later sometimes appears as a single file full of the entire history.

File: Makefile [hunk 6]
```diff
@@ -836,6 +836,12 @@ trwiki_models: \
 trwiki_tuning_reports: \
 	tuning_reports/trwiki.wp10.md
 
+####################### Ukrainian Wikipedia #################################
+datasets/ukwiki.labelings.20200501.json:
+	./utility extract_labelings \
+		/mnt/data/xmldatadumps/public/ukwiki/20200501/ukwiki-20200501-pages-meta-history?*.xml-*.bz2 > $@
+
+
```
Comment (halfak) [inline Makefile:844]: Looks like we'll need some lines to balance the labels, extract text, and extract features.

Comment (chtnnh) [reply Makefile:844]: Working on that right now

**Comment by halfak** [comment]
Looks like this is our lowest fitness model to date, but it's good enough for some testing.  If CI passes, I'd like to merge this, deploy it, and have ukwikipedians test it.

File: Makefile [hunk 7]
```diff
@@ -844,6 +844,88 @@ trwiki_models: \
...
+	  -p 'n_estimators=500' \
+	  -p 'learning_rate=0.01' \
+	  -p 'max_features="log2"' \
+	  -p 'max_depth=7' \
+	  --pop-rate '"I"=0.01798703875' \
+	  --pop-rate '"II"=0.09789710355' \
+	  --pop-rate '"III"=0.60973416214' \
+	  --pop-rate '"IV"=0.26574086557' \
+	  --pop-rate '"ВС"=0.00412565727' \
+	  --pop-rate '"ДС"=0.00984225428' \
```
Comment (halfak) [inline Makefile:918]: We should re-order these in ascending order.  It looks like the order should be:

IV, III, II, I, ВС, ДС

File: articlequality/extractors/ukwiki.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+logger = logging.getLogger(__name__)
+
+NORMALIZED_LABELS = {
+        "ВС": ["ВС", "вс", "Вибрана стаття", "вибрана стаття"],
+        "ДС": ["ДС", "дс", "Добра стаття", "добра стаття"],
+        "I": ["I", "1"],
+        "II": ["II", "2"],
+        "III": ["III", "3"],
+        "IV": ["IV", "4", "Stub", "stub"]
```
Comment (halfak) [inline articlequality/extractors/ukwiki.py:14]: Note the order here corresponds to my previous note.

File: model_info/ukwiki.wp10.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+		label      n         ~I    ~II    ~III    ~IV    ~ВС    ~ДС
+		-------  ---  ---  ----  -----  ------  -----  -----  -----
+		'I'      150  -->    57     34      13      6     18     22
+		'II'     149  -->    30     63      23     21      5      7
+		'III'    150  -->     9     41      51     44      2      3
+		'IV'     150  -->     6     19      42     83      0      0
+		'ВС'     150  -->    13     12       2      0     86     37
+		'ДС'     150  -->    39     11       3      0     39     58
+	rates:
+		              'I'    'II'    'III'    'IV'    'ВС'    'ДС'
```
Comment (halfak) [inline model_info/ukwiki.wp10.md:31]: By changing the order in the cv_train call, it'll make this order appear correctly.

File: Makefile (hunk 7 above)
Comment (chtnnh) [reply Makefile:918]: Do you want me to rebuild the model with the new Makefile to ensure it changes in the model_info as well?

File: articlequality/extractors/ukwiki.py (hunk 8 above)
Comment (chtnnh) [reply articlequality/extractors/ukwiki.py]: Resolved/

File: Makefile (hunk 7 above)
Comment (halfak) [reply Makefile:918]: Yeah.  I think rebuilding should be quick and easy with this small amount of observations.
